CHANGES.md

```diff
@@ -1,4 +1,16 @@
 # Changes
+
+## Unreleased
+* Change `Feature::field` return type from `Result<FieldValue>` to
+  `Result<Option<FieldValue>>`. Fields can be null. Before this change, if a
+  field was null, the value returned was the default value for the underlying
+  type. However, this made it impossible to distinguish between null fields and
+  legitimate values which happen to be default value, for example, an Integer
+  field that is absent (null) from a 0, which can be a valid value. After this
+  change, if a field is null, `None` is returned, rather than the default
+  value.
```

Comment:
Are there likely to be people for whom the old behavior, of getting the default value, is desirable?

If so, after this change, how will they restore that behavior?
Will they need to use snippets like: `.unwrap_or_default(0)` `. unwrap_or_default("")`? or is there a way to say something like "field.default_value()"?

Reply:
(To be clear, I think your change is good overall, just trying to think of any pain we can save a minority of people)

Reply:
Yes, that's right, if they had `dataset.field("some_field")?.into_int()` before, they could do `dataset.field("some_field")?.unwrap_or(FieldValue::Integer(0)).into_int()`. Although it would surprise me what kind of use case there would be where people wouldn't be surprised by nulls turning into `0` or `""`. Probably the more likely case is people aren't aware that nulls silently get turned into real values.

Reply:
> Probably the more likely case is people aren't aware that nulls silently get turned into real values.

Yeah, that seems likely. :+1:

> it would surprise me what kind of use case there would be where people wouldn't be surprised by nulls turning into 0 or "".

I've never personally used this feature, but apparently you can get/set other defaults for fields: https://gdal.org/doxygen/classOGRFieldDefn.html#ac4210fa7c6f10ed090a5558224447cfa

Maybe out of scope for this PR, but since you're thinking about it, what's your opinion on exposing an API leveraging that functionality like:
```
let foo: Result<FieldValue> = dataset.field_default("some_field")?;
```

Then people could exactly mirror the previous behavior with:
```
let foo: Result<FieldValue> = dataset.field("some_field")?.unwrap_or_else(dataset.field_default("some_field")?)
``` 

and maybe offered succinctly as: 
```
let foo: Result<FieldValue> = dataset.field_value_or_default("some_field")?;
```

Reply:
That makes a lot of sense to me. Having a `field_value_or_default` function is clear and descriptive, and exposes useful functionality from OGR.

And since you ask, I was actually starting to think about whether it would be possible to do away with the `FieldValue` type completely and just use generics. I haven't fully fleshed out the details, but I was thinking it should be possible to just make the `field` generic on the return type, so you could just do

```rust
let str_var: String = dataset.field("string_field")?.unwrap();
let int_var: i32 = dataset.field("int_field")?.unwrap();
etc...
```

I think it could be done if we made a `Field` type that mirrors the `OGRFieldDfn` type, then `impl TryFrom<str> for Field<String/i32/etc>`, and then the signature of `field` could be something like

```rust
fn field<T, D>(field_dfn: D) -> Result<Option<T>> where D: Into<Field<T>>
```

The `Field` type could have that `set_default` method on it too.

Reply:
Was there anything else you wanted for this PR? Did you want me to add that example code snippet for getting a default value?

Reply:
i think the PR looks fine. We will wait a few days for other PRs with breaking changes before we release a new version.
Could you add docs to the `field_from_id` method and include that example snippet?

Reply:
Done 🙂